Adjust XAException return codes for better compatibility with XA specification

ochaloup opened the pull request:
This commit fixes error codes which are thrown from PostgreSQL
XAResource jdbc driver to comply better with xa specification.

There are three main adjustments
* XAException.XAER_RMFAIL is used when some connection error happens
  and it is expected that reconnection of RM could occur
* XAException.XAER_NOTA is used when RM does not know anything about the
  provided Xid
* XAException.XAER_PROTO is used when some wrong sequence of method
  calls is invoked

This addresses issues
https://github.com/pgjdbc/pgjdbc/issues/236
https://github.com/pgjdbc/pgjdbc/issues/510
https://github.com/pgjdbc/pgjdbc/issues/683

ochaloup commented:
fixed forgotten checkstyle failures

davecramer commented:
So short of any way to really test this I'm inclined to +1

ochaloup commented:
@davecramer thanks for looking at this!
I will try to run our tests with these changes this week. I will keep you updated.

ochaloup commented:
hi @davecramer , I've got back to this. I re-based the changes and the CI tests seem to pass. Then I built the driver with this changes and run WildFly Narayana integration testsuite with the driver and PostgreSQL 9.3 and 9.4 and all passed.

Would that be ok to merge this PR or is there something more to do/check?

ochaloup commented:
cc to @vlsi and @whitingjr as we were discussing the issue https://github.com/pgjdbc/pgjdbc/issues/510 which this PR covers

davecramer commented:
@ochaloup long week :) 

If nobody else has objections I'm fine with it

whitingjr commented:
@davecramer I've looked over the changes. They look ok.

vlsi commented on pgjdbc/src/main/java/org/postgresql/xa/PGXAConnection.java:
```diff
@@ -395,13 +407,20 @@ public void rollback(Xid xid) throws XAException {
+      if (PSQLState.isConnectionError(ex.getSQLState())) {
+        errorCode = XAException.XAER_RMFAIL;
+      }
+      throw new PGXAException(GT.tr("Error rolling back prepared transaction"), ex, errorCode);
```
@ochaloup , I might be a bit late to the party, however what do you think of printing things like `xid`, `preparedXid`, `committedOrRolledBack` in the error message?

I'm sure, that would be crucial information if somebody receives `""Error rolling back prepared transaction"`.

ochaloup replied:
@vlsi : I agree. Just I was not sure if I can change the format as those information was not part of the messages previously. I'm going to add the proposed info to the messages, push it to the PR and please check that then again.

vlsi commented on pgjdbc/src/main/java/org/postgresql/xa/PGXAConnection.java:
```diff
@@ -442,18 +464,22 @@ private void commitOnePhase(Xid xid) throws XAException {
             XAException.XAER_RMERR);
       }
+      if (!xid.equals(currentXid) || committedOrRolledBack) {
+        throw new PGXAException(GT.tr("One-phase commit with unknown xid "), XAException.XAER_NOTA);
```
space at the end of string

vlsi commented:
@ochaloup , the changes look ok, however, I think it makes sense to add xids (and other relevant state information) to the error messages. Otherwise the error messages are cryptic and it would be hard to analyze.

ochaloup commented:
@vlsi can you please review the current form. I changed some of the error messages, added few debug information printing. Would be fine this way? Does have no harm for translation?

ochaloup commented:
I haven't recognized that I need to run manually `mvn checkstyle:check`. Seems the issue is in the forgotten import. I hope that will be fine this time.

vlsi commented on pgjdbc/src/main/java/org/postgresql/xa/PGXAConnection.java:
```diff
@@ -433,27 +463,34 @@ public void commit(Xid xid, boolean onePhase) throws XAException {
-        // In fact, we don't know if xid is bogus, or if it just wasn't associated with this
-        // connection.
+      if (xid.equals(preparedXid)) {
+        throw new PGXAException(GT.tr("One-phase commit called for xid {0} but connection was prepared with xid {1}",
```
@ochaloup , Did you really mean `xid.equals(preparedXid)` above?
I think the negation is missing. Can you please clarify the logic here?

vlsi commented on pgjdbc/src/main/java/org/postgresql/xa/PGXAConnection.java:
```diff
@@ -195,12 +197,12 @@ public void start(Xid xid, int flags) throws XAException {
     if (flags == TMJOIN) {
       if (state != State.ENDED) {
-        throw new PGXAException(GT.tr("Transaction interleaving not implemented"),
+        throw new PGXAException(GT.tr("Transaction interleaving not implemented. xid={0}, currentXid={1}", xid, currentXid),
```
Just a note: exception message does not reflect the cause. Something like "invalid state" should be used.

ochaloup replied:
@vlsi thank you for your fixes. I haven't managed to understand how the call should be compound correctly (eating humble pie).

In the case of the message - do you expect me to change the message? I can only guess what is the purpose. Do you want to change the message to something like `Start called with transaction join flag while transaction is still active. Transaction interleaving not implemented...`, or...?

whitingjr replied:
Does the exception message still need changing to "invalid state" ? Or something similar ?

whitingjr replied:
"Invalid protocol state requested. Attempted Transaction interleaving is not supported. xid={0}, currentXid={1}"

ochaloup replied:
@whitingjr thank you. I changed the PR to the text you suggested. I only wonder if that's right to say that the protocol state is invalid.
